### PR TITLE
Add current type to the binder when inferring a partial type

### LINKED ISSUE #1819: Variables initialized to None and then to something else don't work with --strict-optional
This works without `--strict-optional`:
```python
def foo() -> str:
    a = None
    a = ''
    return a
```
But with `--strict-optional` we get a complaint like this:
```
error: Incompatible return value type (got "Optional[str]", expected "str")
```
**[UPDATE: the rest is incorrect, please ignore]**
If I annotate the initial assignment again it works without `--strict-optional`:
```python
def foo() -> str:
    a = None  # type: None
    a = ''
    return a
```
But with `--strict-optional` I get this on the first assignment:
```
error: Incompatible types in assignment (expression has type None, variable has type "str")
```
I think both of these ought to work under `--strict-optional`.

I don't know if this requires changes to the binder or elsewhere or both.

### PR DESCRIPTION
**Pull request by ddfisher**
Fix #1819.

### CODE REVIEW

**Comment by gvanrossum** [comment]
I don't understand why there's an `if 0` in the test. There wasn't one in the examples in #1819.

**Comment by ddfisher** [comment]
The test tests two things (which maybe should be separated for clarity).  It tests that `x` is a `str` immediately after assignment, and also that `x` is an `Optional[str]` once the scope of the assignment ends.

File: test-data/unit/check-optional.test [hunk 1]
```diff
@@ -105,9 +105,12 @@ f(None)
 [out]
 main: note: In function "f":
 
-[case testInferOptionalType]
+[case testInferOptionalTypeWithBinder]
```
Comment (gvanrossum) [inline test-data/unit/check-optional.test]: Please split this into two separate tests.

File: test-data/unit/check-optional.test [hunk 2]
```diff
@@ -105,9 +105,12 @@ f(None)
...
 x = None
-x = 1
+if 0:
```
Comment (gvanrossum) [inline test-data/unit/check-optional.test]: Is there a significance to the always-false-ness? Or is this just a convenient condition? I'd much rather see some random bool function used as the condition.

Comment (rwbarton) [inline test-data/unit/check-optional.test]: I like to use `if bool():` for tests like these (and converted a bunch of tests to that style in the binder PR).

Comment (gvanrossum) [inline test-data/unit/check-optional.test]: Brilliant!

**Pull request merged**